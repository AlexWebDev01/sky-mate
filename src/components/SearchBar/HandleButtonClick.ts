export default interface SearchBarInterface {
    handleButtonClick: () => void;
  }