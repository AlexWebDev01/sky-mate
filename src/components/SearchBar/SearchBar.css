.search-bar {
  display: flex;
  width: 285px;
  height: 41px;
  border-radius: 30px;
  background: #FFF;
}
.search-bar input {
  width: 207px;
  border: none;
  outline: none;
  padding: 10px 0 10px 21px;
  border-radius: 30px;
}
.search-bar input::-moz-placeholder {
  color: #CAE7FD;
  font-size: 14px;
  font-weight: 600;
}
.search-bar input::placeholder {
  color: #CAE7FD;
  font-size: 14px;
  font-weight: 600;
}
.search-bar .vertical-line {
  align-self: center;
}
.search-bar button {
  width: auto;
  margin: auto 0 auto 14px;
}/*# sourceMappingURL=SearchBar.css.map */