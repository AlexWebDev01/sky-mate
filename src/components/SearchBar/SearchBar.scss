.search-bar {
  display: flex;
  width: 285px;
  height: 41px;
  border-radius: 30px;
  background: #fff;
  position: absolute;
  top: 40px;
  left: 80px;
  box-shadow: 0px 4px 30px 0px rgba(0, 0, 0, 0.14);

  input {
    width: 228px - 21px;
    border: none;
    outline: none;
    padding: 10px 0 10px 21px;
    border-radius: 30px;
    font-size: 14px;
    font-weight: 600;
  }

  .vertical-line {
    align-self: center;
  }

  button {
    width: auto;
    margin: auto 0 auto 14px;
  }

  .rain::placeholder,
  .snow::placeholder {
    color: var(--light-blue);
  }

  .rain,
  .snow {
    color: var(--blue);
  }

  .sun::placeholder,
  .clear::placeholder {
    color: var(--light-orange);
  }

  .sun,
  .clear {
    color: var(--orange);
  }

  .clouds::placeholder,
  .fog::placeholder,
  .dust::placeholder,
  .mist::placeholder {
    color: var(--light-green);
  }

  .clouds,
  .fog,
  .dust,
  .mist {
    color: var(--green);
  }
}
