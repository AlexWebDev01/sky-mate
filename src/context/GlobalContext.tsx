import { useState, createContext, useContext, FunctionComponent } from "react";
import {
  Coordinates,
  GlobalContext,
  GlobalProviderProps,
} from "./GlobalContext.interface";

const GlobalContext = createContext<GlobalContext | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error("useWords must be used within a WordsProvider");
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);
  const [weatherData, setWeatherData] = useState<object | null>(null);
  const [location, setLocation] = useState<string | null>(null);
  const [pageStyle, setPageStyle] = useState<string>("");

  const sharedState: GlobalContext = {
    coordinates,
    setCoordinates,
    weatherData,
    setWeatherData,
    location,
    setLocation,
    pageStyle,
    setPageStyle,
  };

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
