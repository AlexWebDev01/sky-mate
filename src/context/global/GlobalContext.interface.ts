import { ReactNode } from 'react';
import {
  ClothesAdviceWithTemp,
  WeatherConditions,
} from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import {
  Coordinates,
  WeatherData,
} from '@api/fetchWeatherData/fetchWeatherData.interface';

export interface GlobalContext {
  state: GlobalState;
  fetchData: () => Promise<void>;
  handleSearch: (location: string) => Promise<void>;
}

export interface GlobalState {
  isLoading: boolean;
  coordinates: Coordinates | null;
  location: string | null;
  weatherData: WeatherData | null;
  pageStyle: WeatherConditions | '';
  clothesAdvice: ClothesAdviceWithTemp | null;
}

export interface GlobalProviderProps {
  children: ReactNode;
}
