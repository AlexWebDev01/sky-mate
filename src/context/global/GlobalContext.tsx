import {
  useState,
  createContext,
  useContext,
  FunctionComponent,
  useCallback,
  useMemo,
} from 'react';
import {
  GlobalContext as GlobalContextInterface,
  GlobalProviderProps,
  GlobalState,
} from './GlobalContext.interface';

import { fetchLocationData } from '@api/fetchLocationData/fetchLocationData';
import { fetchUserLocatioByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP';
import { separateCoordinates } from '@shared/helpers/date/date';
import { fetchWeatherData } from '@api/fetchWeatherData/fetchWeatherData';
import { calculateClothesAdvice } from '@shared/helpers/clothesAlgorithm';

import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import { LocationDataByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP.interface';
import { LocationData } from '@api/fetchLocationData/fetchLocationData.interface';
import {
  Units,
  Coordinates,
} from '@api/fetchWeatherData/fetchWeatherData.interface';

const GlobalContext = createContext<GlobalContextInterface | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Global context should be used in GlobalProvider!');
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [state, setState] = useState<GlobalState>({
    isLoading: true,
    coordinates: null,
    location: null,
    weatherData: null,
    pageStyle: '',
    clothesAdvice: null,
  });
  const [error, setError] = useState<string>('');

  const getWeatherData = useCallback(async (coordinates: Coordinates) => {
    const data = await fetchWeatherData(coordinates, Units.metric);
    const weatherCondition =
      data.daily[0].weather[0].main.toLowerCase() as WeatherConditions;
    const averageTemp = data.daily[0].temp.day;
    const clothesAdvice = calculateClothesAdvice(weatherCondition, averageTemp);
    return {
      weatherData: data,
      clothesAdvice,
      pageStyle: weatherCondition,
    };
  }, []);

  const fetchData = useCallback(async () => {
    console.group('FETCH DATA');
    console.info('Started initial data fetching');

    setState((prevState) => ({ ...prevState, isLoading: true }));

    try {
      const locationData: LocationDataByIP = await fetchUserLocatioByIP();
      const coordinates = separateCoordinates(locationData.loc);

      const { weatherData, clothesAdvice, pageStyle } =
        await getWeatherData(coordinates);
      setState((prevState) => ({
        ...prevState,
        weatherData,
        clothesAdvice,
        pageStyle,
        coordinates,
        location: locationData.city,
        isLoading: false,
      }));

      console.info('Fetched & calculated data: ', {
        weatherData,
        clothesAdvice,
        pageStyle,
        city: locationData.city,
      });
      console.info('Finished initial data fetching');
      console.groupEnd();
    } catch (error) {
      if (error instanceof Error) {
        setError('Oops, unexpected error!');
      }

      console.error('Failed to fetch initial data:', error);
    }
  }, [getWeatherData]);

  const handleSearch = useCallback(
    async (location: string) => {
      try {
        console.group('HANDLE SEARCH');
        console.info('Started fetching data for location:', location);
        const locationData: LocationData = await fetchLocationData(location);
        const { name, lat, lon } = locationData;

        const { weatherData, clothesAdvice, pageStyle } = await getWeatherData({
          lat,
          lon,
        });

        setState((prevState) => ({
          ...prevState,
          coordinates: { lat, lon },
          location: name,
          weatherData,
          clothesAdvice,
          pageStyle,
        }));
        console.info('Fetched & calculated data: ', {
          weatherData,
          clothesAdvice,
          pageStyle,
          city: name,
        });
        console.info('Finished fetching data for location:', location);
      } catch (error) {
        if (error instanceof Error && error.message === 'Location not found') {
          setError('Location not found');
        } else if (error instanceof Error) {
          setError('Oops, unexpected error!');
        }

        console.error('Error fetching data for location:', error);
      }
    },
    [getWeatherData]
  );

  const sharedState = useMemo<GlobalContextInterface>(
    () => ({
      state,
      error,
      setError,
      fetchData,
      handleSearch,
    }),
    [state, error, setError, fetchData, handleSearch]
  );

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
