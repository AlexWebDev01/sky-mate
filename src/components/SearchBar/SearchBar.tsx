import { useState } from 'react';

import { useGlobalContext } from '@context/GlobalContext';
import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';

import './SearchBar.css';

const SearchBar = () => {
  const { state, handleSearch } = useGlobalContext();
  const [inputValue, setInputValue] = useState('');
  const { pageStyle } = state;

  const performSearch = () => {
    const trimmedValue = inputValue.trim();
    if (trimmedValue) {
      handleSearch(trimmedValue);
      setInputValue('');
    }
  };

  const onClickHandler = () => {
    performSearch();
  };

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setInputValue(value);
  };

  const onKeyPressHandler = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter') {
      performSearch();
    }
  };

  const color: string =
    pageStyle === WeatherConditions.snow || pageStyle === WeatherConditions.rain
      ? 'var(--light-blue)'
      : pageStyle === WeatherConditions.clear ||
          pageStyle === WeatherConditions.sun
        ? 'var(--light-orange)'
        : pageStyle === WeatherConditions.clouds ||
            pageStyle === WeatherConditions.fog ||
            pageStyle === WeatherConditions.dust ||
            pageStyle === WeatherConditions.mist
          ? 'var(--light-green)'
          : '';

  return (
    <div className='search-bar'>
      <input
        placeholder='Search'
        className={pageStyle}
        onKeyDown={onKeyPressHandler}
        onChange={handleChange}
        value={inputValue}
        autoFocus
      />
      <svg
        className='vertical-line'
        width='2'
        height='32'
        viewBox='0 0 2 32'
        fill='none'
        xmlns='http://www.w3.org/2000/svg'
      >
        <path id='Vector 10' d='M1 0L0.999999 32' stroke={color} />
      </svg>
      <button onClick={onClickHandler}>
        <svg
          width='25'
          height='25'
          viewBox='0 0 25 25'
          fill='none'
          xmlns='http://www.w3.org/2000/svg'
        >
          <path
            id='Vector'
            d='M24 24L18.4507 18.4507M18.4507 18.4507C19.3999 17.5014 20.1529 16.3745 20.6666 15.1343C21.1803 13.8941 21.4447 12.5648 21.4447 11.2224C21.4447 9.87995 21.1803 8.55067 20.6666 7.31044C20.1529 6.0702 19.3999 4.9433 18.4507 3.99406C17.5014 3.04483 16.3745 2.29185 15.1343 1.77813C13.8941 1.26441 12.5648 1 11.2224 1C9.87995 1 8.55067 1.26441 7.31044 1.77813C6.0702 2.29185 4.9433 3.04483 3.99406 3.99406C2.077 5.91113 1 8.51123 1 11.2224C1 13.9335 2.077 16.5336 3.99406 18.4507C5.91113 20.3677 8.51123 21.4447 11.2224 21.4447C13.9335 21.4447 16.5336 20.3677 18.4507 18.4507Z'
            stroke={color}
            strokeLinecap='round'
            strokeLinejoin='round'
          />
        </svg>
      </button>
    </div>
  );
};

export default SearchBar;
