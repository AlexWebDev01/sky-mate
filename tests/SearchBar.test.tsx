import '@testing-library/jest-dom';
import { render, screen, fireEvent, act } from '@testing-library/react';
import SearchBar from '@components/SearchBar/SearchBar';
import { useGlobalContext } from '@context/global/GlobalContext';
import { GlobalContext } from '@context/global/GlobalContext.interface';

jest.mock('@shared/assets/icons/vertical-line.svg', () => {
  return () => <svg data-testid='mocked-vertical-line'></svg>;
});

jest.mock('@shared/assets/icons/search-icon.svg', () => {
  return (props: { onClick: () => void }) => (
    <svg data-testid='mocked-search-icon' onClick={props.onClick}></svg>
  );
});

jest.mock('@context/global/GlobalContext', () => ({
  useGlobalContext: jest.fn(),
}));

const mockUseGlobalContext = useGlobalContext as jest.MockedFunction<
  typeof useGlobalContext
>;

describe('SearchBar Component', () => {
  const mockContextValue: GlobalContext = {
    state: {
      pageStyle: '',
      isLoading: true,
      coordinates: null,
      location: null,
      weatherData: null,
      clothesAdvice: null,
    },
    handleSearch: jest.fn(),
    fetchData: jest.fn(),
  };

  beforeEach(() => {
    mockUseGlobalContext.mockReturnValue(mockContextValue);
  });

  test('renders the component', () => {
    render(<SearchBar />);
    const inputElement = screen.getByPlaceholderText('Search');
    const verticalLineElement = screen.getByTestId('mocked-vertical-line');
    const searchIconElement = screen.getByTestId('mocked-search-icon');

    expect(inputElement).toBeInTheDocument();
    expect(verticalLineElement).toBeInTheDocument();
    expect(searchIconElement).toBeInTheDocument();
  });

  test('updates input value on change', () => {
    render(<SearchBar />);
    const inputElement = screen.getByPlaceholderText('Search');
    fireEvent.change(inputElement, { target: { value: 'test query' } });

    expect(inputElement).toHaveValue('test query');
  });

  test('performs search on enter key press', () => {
    render(<SearchBar />);
    const inputElement = screen.getByPlaceholderText('Search');
    fireEvent.change(inputElement, { target: { value: 'test query' } });

    fireEvent.keyDown(inputElement, { key: 'Enter', code: 'Enter' });

    expect(mockContextValue.handleSearch).toHaveBeenCalledWith('test query');
    expect(inputElement).toHaveValue('');
  });

  test('performs search on search icon click', () => {
    render(<SearchBar />);
    const inputElement = screen.getByPlaceholderText('Search');
    fireEvent.change(inputElement, { target: { value: 'test query' } });
    const searchIconElement = screen.getByTestId('mocked-search-icon');

    act(() => {
      fireEvent.click(searchIconElement);
    });

    expect(mockContextValue.handleSearch).toHaveBeenCalledWith('test query');
    expect(inputElement).toHaveValue('');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<SearchBar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
