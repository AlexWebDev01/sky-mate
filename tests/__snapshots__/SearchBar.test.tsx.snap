// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="search-bar "
  >
    <input
      placeholder="Search"
      value=""
    />
    <svg
      data-testid="mocked-vertical-line"
    />
    <svg
      data-testid="mocked-search-icon"
    />
  </div>
</DocumentFragment>
`;
